### ISSUE TITLE
[localize] Support localization of dependencies

### ISSUE THREAD

**Issue by aomarks** [Polymer/lit-html#1661]
#### Problem

Currently, Localize only considers the immediate source files of your TypeScript project. If you have a dependency or transitive dependency that also needs localization, then there is currently no obvious way to handle it. Transform mode doesn't work at all, because we don't transform dependencies. Runtime mode *could* work, but it would require manually merging the XLIFF files of each dependency in order to generate a complete locale module.

#### Proposal

- During analysis, transitively walk into dependencies by following import statements and resolving to `node_modules/` packages. This will require building a new TypeScript program for each dependency.

- For each dependency package, look for a `lit-localizations.json` file or look up the `{"lit": {"localizations": "..."}}` property in `package.json` (names TBD). This is an optional file that allows package authors to publish pre-translated strings for their package. Automatically merge all of these files into our own translations during analysis.

- If a package doesn't contain existing translations, or doesn't include a language we need, then we'll adopt that dependency's messages as our own, and emit translation requests into our own XLIFF files.

- In runtime mode, the above should be sufficient. We'll end up with generated locale modules that contain our top-level translations, plus the translations of every transitive dependency (either provided by them, or adopted and translated by us).

- In transform mode, we'll add a copy of `node_modules/` to every locale output folder, and directly transform those module's sources.

**Comment by aomarks** [comment]
Had a discussion with @jorenbroekema about localization in dependencies, and also sharding:

### Requirements

#### High level requirements:

1. A library can bundle translations in its package, and consumers can use them.
2. A user can add new locales on top of the bundled ones
3. A user can override bundled translations for specific messages.
4. Translations can be sharded, so that you don't have to load all messages for a locale at once, e.g. so that you can load your core app messages first, and then defer loading translations for some specific component until you actually need them.

#### Zero-build requirements:

5. A user can use bundled translations without a build step.
6. A user can add new locales without a build step
7. A user can override bundled translations for specific messages without a build step.

#### Things we want to preserve:

8. Any new APIs are statically analyzable, so that transform mode always works. The end-user should always be able to do a full transformed build, where localized messages are baked right into the source, and all runtime localization code is shaken out.
9. Sharding is completely configurable by the end-user. Maybe there is a default package or module sharding, but the user should be able to precisely declare how they want translations to be sharded (just like you can do in rollup).

## Proposal

We settled on the idea of API something like this:

#### Component library
```ts
import {createNamespace} from `@lit/localize`;
export const localization = createNamespace({
  translationsDir: './translations/',
  sourceLocale: "nl",
  targetLocales: ["en-GB", ...]
});
const {msg} = localization;
msg("My namespaced translation", {id: "my-id"})
```

#### Consumer

```ts
import {localization} from "dependency";
localization.replaceTranslation("en-GB", {
  "my-id", "A better translation",
  "another-id":, ...
});
localization.overlayTranslations("en-GB", "./my-extra-translations-dir/")
```

- The new `createNamespace` API allows a module to declare a set of local translations. A path to a directory is specified, e.g. `"./translations"`, which is interpreted as relative to the `import.meta.url`. It returns an object with a namespace-scoped `msg` function, plus some functions that consumers can use to modify/augment the translations if needed.

- A user who imports the dependency can now use the provided translations _without running any build step themselves_, because lit localize already knows where to find the bundled translations for each namespace.

- The component library *exports* the namespace object, which allows users to augment the translations. E.g. if they need to fix a bad translation, or add a new locale.

### Transform/bundling config

If the user wants more control over the way translations are sharded, they can install `@lit/localize-tools` and customize the configuration.

Since everything is statically analyzable, we can still do a full transform build, removing all runtime code as usual. Bundled messages are merged with any user augmentations.

Once we have a [runtime build](https://github.com/Polymer/lit-html/issues/1469), we can also do configurable bundling. Some configurations a user might want:

- One one module per locale (the current output)
- One mega-module for all locales (e.g. a tiny app with very few locales that wants to bundle everything in JS)
- A shard per view in a SPA

The sharding config API might look something like:

```js
export default {
  messageModulePath: (locale, messageId, sourcePath) => {
    if (sourcePath.startsWith('/core/')) {
      return `locales/core/${locale}.js`
    }
    return `locales/everything-else/${locale}.js`,
  }
}
```

Messages would then be arranged into modules according to this function, and the transformed application output would include a manifest that tells it the location of any given message, so that it knows which module to fetch.
 
### Sharing a namespace

If a package wants to share a namespace between two modules, it can do this:

##### ./localization.js
```ts
import {createNamespace} from `@lit/localize`;
export const localization = createNamespace({
  translationsDir: './translations/',
  sourceLocale: "nl",
  targetLocales: ["en-GB", ...]
});
```

##### ./component-a.js
```ts
import {localization} from './localization.js';
const {msg} = localization;
msg("Component A namespaced message");
```

##### ./component-b.js
```ts
import {localization} from './localization.js';
const {msg} = localization;
msg("Component B namespaced message");
```

**Comment by jorenbroekema** [comment]
Hey, thanks a lot for doing the write up on our talks on Slack! 

Here's a link to the requirements that I put in a markdown file https://github.com/jorenbroekema/lit-localize-test/blob/main/requirements.md 

It outlines most of our requirements, lacks prioritisation at the moment and some of the "override/edit" localization namespaces are quite similar. I tried marking the requirements with whether `@lit/localize` already supports the use cases or makes it trivial to add on our end. A red cross basically means there's no clear way to accomplish it with the current `@lit/localize` but with the proposals above I think it would cover everything, at least as far as I can see :).

I really like the sharding strategy that scopes on a module level, as it prevents name clashes and allows scoping on a level that should definitely be sufficient and not overly specific (e.g. scoping per lit component), with the added benefit of easily sharing scopes by using ES Modules syntax.

Static analysis as a requirement makes flexibility a little trickier, I think we managed to find a reasonable solutions that preserve this though while covering the requirements we have 👍

**Comment by lesqui23** [comment]
Are there still considerations to implement the proposal? I'm a big fan of the lit framework and have created a library of components for fast app development. For this I come across the exact topic regarding localization. Especially for the aggregations with more functionality.

My workaround at the moment is to import a wrapper within localisation.js (wich I hope is a small usefull implementation):
`
import {templates as dependenyComponentTemplates} from 'dependency-component/src/generated/locale/en.js';
import {templates as originTemplaes} from '../generated/locale/en.js';

export const templates = {
  ...originTemplaes,
  ...formTemplates
};`

**Comment by aomarks** [comment]
> Are there still considerations to implement the proposal?

Yes! This is still an important priority for Lit Localize. Sorry there hasn't been a lot of progress on this topic since it was filed. I'm expecting to start working on this again in the next month or so.

**Comment by lesqui23** [comment]
Thanks, great news! I looking forward to integrate the upcoming changes into my localization tool .

**Comment by GustavPS** [comment]
Any updates on this? I would like to have the same translation key (ID) in different files and they would generate unique translations. I.e for example I have "key1" used in msg() in two different files. I would like them to have two different translations and I think this proposal is the way to go for that.

**Comment by aomarks** [comment]
> Any updates on this?

Nothing new to report, other than to report it is still high on my priority list. There are currently a number of other things vying for time, unfortunately :/

> I would like to have the same translation key (ID) in different files and they would generate unique translations. I.e for example I have "key1" used in msg() in two different files. I would like them to have two different translations and I think this proposal is the way to go for that.

Are you using automatic or manually specified IDs? When you say `"key1"`, do you mean you are specifying that as the id, or that is the actual content of the message?

**Comment by GustavPS** [comment]
> Are you using automatic or manually specified IDs? When you say `"key1"`, do you mean you are specifying that as the id, or that is the actual content of the message?

As the actual content of the message.
Our repo(s) have a lot of components. Our translation backend save translations on a component basis. So let's say I do a
`msg(weather)` in component A and a `msg(weather)` in component B. Those could have 2 totally different meaning depending on which component they are in. And from what I've read a msg with the same string (weather in this example) will give the same ID even if they are in two different files (components in this case). Please correct me if I'm wrong though 😄

So basically what I want to achieve is to have the same msg (or id) in two different files and be able to set different translations on both of them

**Comment by aomarks** [comment]
> As the actual content of the message. Our repo(s) have a lot of components. Our translation backend save translations on a component basis. So let's say I do a `msg(weather)` in component A and a `msg(weather)` in component B. Those could have 2 totally different meaning depending on which component they are in. And from what I've read a msg with the same string (weather in this example) will give the same ID even if they are in two different files (components in this case). Please correct me if I'm wrong though smile
> 
> So basically what I want to achieve is to have the same msg (or id) in two different files and be able to set different translations on both of them

Gotcha. That usage is cutting against the grain of how the library is intended to be used currently, though I can see how you made it work :)

The intended usage pattern for Lit Localize is that you write _actual_ message contents in your templates using your preferred authoring language, just as you would do when writing for a single locale, and then generate the other locales. I'm curious to hear why that pattern isn't working for you.

This issue could maybe be helpful for you, because it would let you associate each message with a namespace, and then you could assign an id to the namespace. That's similar to providing a distinct id for each message, though.

**Comment by GustavPS** [comment]
> Gotcha. That usage is cutting against the grain of how the library is intended to be used currently, though I can see how you made it work :)
> 
> The intended usage pattern for Lit Localize is that you write _actual_ message contents in your templates using your preferred authoring language, just as you would do when writing for a single locale, and then generate the other locales. I'm curious to hear why that pattern isn't working for you.
> 
> This issue could maybe be helpful for you, because it would let you associate each message with a namespace, and then you could assign an id to the namespace. That's similar to providing a distinct id for each message, though.

It's not working for us because we can't really change our old translation backend which contains a lot of duplicate keys that would be used in the `msg` call. I'm doing a POC for changing from Polymer to Lit and hoped that we could use the official Lit localization system, but maybe we have to build our own like we did for Polymer 😄 But hopefully the namespacing could be an option for us when it's released

**Comment by ca-d** [comment]
Is anyone currently still pushing this issue forward? If not, is there something we could do to help?

As authors of both a component library and an app which consumes it, we would very much appreciate handling our translations using an API like the one proposed here.

**Comment by aomarks** [comment]
> Is anyone currently still pushing this issue forward? If not, is there something we could do to help?
> 
> As authors of both a component library and an app which consumes it, we would very much appreciate handling our translations using an API like the one proposed here.

Sorry for the slow progress on this. This has been on my plate and I still consider it the most important thing for Localize, but other priorities have taken my attention. Will try to give an update soon.

**Comment by guillemcordoba** [comment]
Any new updates @aomarks ? I'm starting to need this...

**Comment by enqueue** [comment]
Is it perhaps possible to break this problem down into smaller tasks? Perhaps support for namespaces / bundles similar to lion, as suggested by @jorenbroekema,  and support for overriding "third-party" localizations?

**Comment by ca-d** [comment]
> This has been on my plate

@aomarks is there anyone else whose plate this could be put onto? I'd really appreciate some progress here, otherwise we'll have to revert to `lit-translate` and our custom json format for translations, which some of our stakeholders (who love XLIFF) would consider a real shame.

**Comment by jpzwarte** [comment]
I just managed to get runtime translation working with translations in my app and in the design system by using the following code:

```js
const { setLocale } = configureLocalization({
  sourceLocale: 'en',
  targetLocales: ['nl'],
  // Step 1: make this an async method
  loadLocale: async locale => {
    // Step 2: Load both the templates of the app and the library (assuming the library also publishes its templates)
    const [{ templates }, { templates: dnaTemplates }] = await Promise.all([
      import(`./src/locales/${locale}.js`),
      import(`./node_modules/@dna/core/locales/${locale}.js`)
    ]);

    // Step 3: Merge the templates and return them as one
    return { templates: { ...templates, ...dnaTemplates }};
  }
});
```

**Comment by gmzacharydovel** [comment]
I'm actually in the middle now if looking into how to split up translations by page.  Part of the problem is that lit doesn't include the source file context when extracting, so it can make it harder to figure out how to split things up.

But yes, @jpzwarte ideally it would just be a shallow merge of the json like you have there.  In fact, that is probably closest to the correct answer in my opinion.

@aomarks I'm open to looking into this after the other localization tickets if you want me contributing
